enterprise/decorators.py

```diff
@@ -0,0 +1,175 @@
+    we need to be able to wrap the function with something.
+    """
+    return func
+
+
+def enterprise_login_required(view):
+    """
+    View decorator for allowing authenticated user with valid enterprise UUID.
+
+    This decorator requires enterprise identifier as a parameter
+    `enterprise_uuid`.
+
+    This decorator will throw 404 if no kwarg `enterprise_uuid` is provided to
+    the decorated view .
+
+    If there is no enterprise in database against the kwarg `enterprise_uuid`
+    or if the user is not authenticated then it will redirect the user to the
+    enterprise-linked SSO login page.
+
+    Usage::
+        @enterprise_login_required()
+        def my_view(request, enterprise_uuid):
+            # Some functionality ...
+
+        OR
+
+        class MyView(View):
+            ...
+            @method_decorator(enterprise_login_required)
+            def get(self, request, enterprise_uuid):
+                # Some functionality ...
+
+    """
+    @wraps(view)
+    def wrapper(request, *args, **kwargs):
+        """
+        Wrap the decorator.
+        """
+        if 'enterprise_uuid' not in kwargs:
+            raise Http404
+
+        enterprise_uuid = kwargs['enterprise_uuid']
+        enterprise_customer = get_enterprise_customer_or_404(enterprise_uuid)
+
+        # Now verify if the user is logged in. If user is not logged in then
+        # send the user to the login screen to sign in with an
+        # Enterprise-linked IdP and the pipeline will get them back here.
+        if not request.user.is_authenticated():
+            next_url = '{current_url}?{query_string}'.format(
+                current_url=quote(request.get_full_path()),
+                query_string=urlencode({'tpa_hint': enterprise_customer.identity_provider})
+            )
+            return redirect(
+                '{login_url}?{params}'.format(
+                    login_url='/login',
+                    params=urlencode(
+                        {'next': next_url}
+                    )
+                )
+            )
+
+        # Otherwise, they can proceed to the original view.
+        return view(request, *args, **kwargs)
+
+    return wrapper
+
+
+def force_fresh_session(view):
+    """
+    View decorator which terminates stale TPA sessions.
+
+    This decorator forces the user to obtain a new session
+    the first time they access the decorated view. This prevents
+    TPA-authenticated users from hijacking the session of another
+    user who may have been previously logged in using the same
+    browser window.
+
+    Usage::
+        @force_fresh_session()
+        def my_view(request, enterprise_uuid):
+            # Some functionality ...
+
+        OR
+
+        class MyView(View):
+            ...
+            @method_decorator(force_fresh_session)
+            def get(self, request, enterprise_uuid):
+                # Some functionality ...
+
```

Comment:
Is it true that this decorator wouldn't really make sense to use without the login required decorator? Would that be worth noting?